pkg/restart/restart.go

```diff
@@ -0,0 +1,166 @@
+package restart
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+var (
+	restartNeededError  = errors.New("restart needed called")
+	alreadyRunningError = errors.New("controller manager already running")
+
+	_ Restarter = (*StartManager)(nil)
+)
+
+type InitializeManagerFunc func() (manager.Manager, error)
+
+type Restarter interface {
+	RestartNeeded()
+}
+
+type StartManager struct {
+	InitializeManager InitializeManagerFunc
+
+	running        bool
+	lock           sync.Mutex
+	runningContext context.Context
+	cancelFunc     context.CancelCauseFunc
+}
+
+func debug(message string) {
+	// fmt.Println(message)
+}
+
+func (s *StartManager) Start(ctx context.Context) error {
+	for {
+		debug("Start: start loop enter")
+		startErr, finished := s.doStart(ctx)
+
+		if s.runningContext == nil {
+			// this actually means that the start failed, but we need a non-nil value for the runningContext
+			// so that we can select on it below
+			s.runningContext = context.TODO()
+		}
+
+		debug("Start: selecting on cancellation and error channels")
+		select {
+		case <-s.runningContext.Done():
+			{
+				debug("Start: context cancelled, waiting for finish signal")
+				// The context is cancelled so the manager is shutting down. Let's wait for it to actually finish running before allowing
+				// ourselves to continue and eventually allowing another round of manager run.
+				<-finished
+
+				debug("Start: context cancelled, finish signal received")
+
+				// the manager is not running anymore, now we can be sure of it. Let's record that fact in our state to keep the records
+				// straight.
+				debug("Start: context cancelled, about to set running=false")
+				s.lock.Lock()
+				debug("Start: locked")
+				s.running = false
+				s.lock.Unlock()
+				debug("Start: unlocked")
+				debug("Start: context cancelled, running=false is set")
+
+				// now, let's see what caused the cancellation
+
+				if !errors.Is(context.Cause(s.runningContext), restartNeededError) {
```

Comment:
checking the cause of the cancel function and comparing it is a bit weird to me - it reminds me catching exceptions in Java. Not sure if this is a best practice in Go though.

Instead of using "CancelCause Context" you could use the standard "Cancel context" and add another field to the `StartManager` stating if the restart was triggered or not (in the same way as you do it with the `running` field. And have another case when the root context itself:
```go
		case <-ctx.Done():
			debug("Start: context cancelled, cancellation from above")
			<-finished
			debug("Start: quitting")
```